x-pack/plugins/security_solution/public/resolver/models/node_data.ts

```diff
@@ -0,0 +1,148 @@
+  const originalData = storedNodeInfo ?? [];
+  const newData = new Map<string, NodeData>([...originalData]);
+  newData.delete(nodeID);
+  return newData;
+}
+
+/**
+ * Creates a copy of the node data map and updates it with the data returned by the server. If the server did not return
+ * data for a particular ID we will determine whether no data exists for that ID or if the server reached the limit we
+ * requested by using the reachedLimit flag.
+ *
+ * @param storedNodeInfo the node data from state
+ * @param receivedNodes the events grouped by ID that the server returned
+ * @param requestedNodes the IDs that we requested the server find events for
+ * @param reachedLimit a flag indicating whether the server returned the same number of events we requested
+ */
+export function updateWithReceivedNodes({
+  storedNodeInfo,
+  receivedNodes,
+  requestedNodes,
+  reachedLimit,
+}: {
+  storedNodeInfo: IDToNodeInfo | undefined;
+  receivedNodes: Map<string, FetchedNodeData>;
+  requestedNodes: Set<string>;
+  reachedLimit: boolean;
+}): IDToNodeInfo {
+  const copiedMap = new Map<string, NodeData>([...(storedNodeInfo ?? [])]);
+
+  for (const id of requestedNodes.values()) {
+    // If the server returned the same number of events that we requested it's possible
+    // that we won't have node data for each of the IDs. So we'll want to remove the ID's
+    // from the map that we don't have node data for
+    if (!receivedNodes.has(id)) {
+      if (reachedLimit) {
+        copiedMap.delete(id);
+      } else {
+        // if we didn't reach the limit but we didn't receive any node data for a particular ID
+        // then that means Elasticsearch does not have any node data for that ID.
+        copiedMap.set(id, { events: [], status: 'running' });
+      }
+    }
+  }
+
+  // for the nodes we got results for, create a new array with the contents of those events
+  for (const [id, info] of receivedNodes.entries()) {
+    copiedMap.set(id, {
+      events: [...info.events],
+      status: info.terminated ? 'terminated' : 'running',
+    });
+  }
+
+  return copiedMap;
+}
+
+/**
+ * Checks to see if a set of node IDs exists in state's node data.
+ *
+ * @param baseNodes the node data map from state
+ * @param nodesToCheck a set of node IDs
+ */
+export function idsNotInBase(
+  baseNodes: IDToNodeInfo | undefined,
+  nodesToCheck: Set<string>
+): Set<string> {
+  const result = new Set<string>();
+
+  for (const id of nodesToCheck.values()) {
+    if (!baseNodes || !baseNodes.has(id)) {
+      result.add(id);
+    }
+  }
+
+  return result;
+}
+
+/**
+ * This is used for displaying information in the node panel mainly and we should be able to remove it eventually in
+ * favor of showing all the node data associated with a node in the tree.
+ *
+ * @param data node data for a specific node ID
+ * @returns the first event or undefined if the node data passed in was undefined
+ */
+export function firstEvent(data: NodeData | undefined) {
```

Comment:
Could you add a return type to this function? I think it could be helpful, especially since this function is exported.